### PR TITLE
New sid plugin

### PR DESCRIPTION
**Pull request by lemikev**
plugin used to generate .sid file as defined in https://datatracker.ietf.org/doc/draft-somaraju-core-sid/

Usage:

> pyang --sid-help

Structure IDentifier (SID) are used to map YANG definitions to
CBOR encoding. These SIDs can be automatically generated
for a YANG module using the pyang sid plugin.

   pyang --generate-sid-file entry-point:size <yang-module>

For example:
   pyang --generate-sid-file 20000:100 toaster@2009-11-20.yang

The name of the .sid file generated is:

   <module-name>@<module-revision>.sid

Each time new items(s) are added to a YANG module by the
introduction of a new revision of this module, its included
sub-module(s) or imported module(s), the .sid file need to be
updated. This is done by providing the name of the previous
.sid file as argument.

   pyang --update-sid-file <file-name> <yang-module>

For example:
   pyang --update-sid-file toaster@2009-11-20.sid toaster@2009-12-28.yang

The --check-sid-file option can be used at any time to verify
if the .sid file need to be updated.

   pyang --check-sid-file <file-name> <yang-module>

The --list_sid option can be included before any of the previous
option to obtains the list of SIDs assigned or validated.

For example:
   pyang --list-sid --generate-sid-file 20000:100 toaster@2009-11-20.yang

### CODE REVIEW

**Comment by mbj4668** [comment]
Hi,

Some general comments here, and some specific inline in the code.

Please try to format your file to fit 80 characters per line.

I would prefer the following options:

  -f sid --sid-range <entry-point:size>
  -f sid --sid-update-from <old-sid file>

Possibly add the option --sid-list or --sid-verbose or something to show the listing in plain text.  But OTOH, it seems a separate tool would be better:  sidlist <sid file>.

Please also document your options in doc/pyang.1.dbk.

And last, please add some tests, create a new directory test/test_sid; do positive and negative testing (e.g., I tried to reference a non-existing file with --update-sid-file, and it gave me a stacktrace)

File: pyang/plugins/sid.py [hunk 1]
```diff
@@ -0,0 +1,566 @@
+"""sid plugin
+
+Plugin used to generate or update .sid files.
```
Comment (mbj4668) [inline pyang/plugins/sid.py]: Please include a reference to the relevant specs.

File: pyang/plugins/sid.py [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+   pyang --list-sid --generate-sid-file 20000:100 toaster@2009-11-20.yang
+"""
+)
+
+############################################################
+class SidFileError(Exception):
+    def __init__(self, msg=""):
+        self.msg = msg
+
+class SidParcingError(Exception):
```
Comment (mbj4668) [inline pyang/plugins/sid.py]: This should be SidParsingError

File: pyang/plugins/sid.py [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+        components = range.split(':')
+        if len(components) != 2 or not re.match(r'\d+:\d+', range):
+            raise SidParcingError("invalid range in argument, must be '<entry-point>:<size>'.")
+
+        self.content = OrderedDict([('assignment-ranges', [])])
+        self.content['assignment-ranges'].append(OrderedDict([('entry-point', int(components[0])), ('size', int(components[1]))]))
+
+    ########################################################
+    # Retrieve the module revision from the pyang context
+    def get_module_revision(self, module):
```
Comment (mbj4668) [inline pyang/plugins/sid.py]: Please use util.get_latest_revision(module)

File: pyang/plugins/sid.py [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+        if 'module-name' not in self.content or self.module_name != self.content['module-name']:
+            self.content['module-name'] = self.module_name
+            if self.check_consistency == True:
+                print("ERROR, Mismatch between the module name defined in the .sid file and the .yang file.")
+                self.is_consistent = False
+
+        if 'module-revision' not in self.content or self.module_revision != self.content['module-revision']:
+            self.content['module-revision'] = self.module_revision
+            if self.check_consistency == True:
+                print("ERROR, Mismatch between the module revision defined in the .sid file and the .yang file.")
```
Comment (mbj4668) [inline pyang/plugins/sid.py]: Rather than printing errors directly, the idea is to register a plugin-specific error code, and set it.

Also, in general there is an error EmitError that should be used if possible to generate errors from the emit process (assumes you are using -f sid).

**Comment by gruber6121** [comment]
Hi,
It is possible for someone to provide me the steps to install the plugin?
Many thanks in advance.

**Comment by lemikev** [comment]
Hi

To install this plugin, you simply need to copy the file "sid.py" in "pyang\pyang\plugins".
Than, you should be able to request help for it.

> python.exe pyang --sid-help
The latest version is in attachment.

I plan to complete the integration of this plugin in github prior to IETF99.

Regards,

[cid:image001.jpg@01C868D8.BF0BB7E0]

Michel Veillette
System Architecture Director
Trilliant Inc.
Tel: 450-531-3109
https://trilliantinc.com/

**Comment by gruber6121** [comment]
Hi,
The file "sid.py" is in "pyang\pyang\plugins"
Unfortunately, after  executing "python.exe pyang --sid-help" the output is "python.exe: command not found"

**Comment by lemikev** [comment]
Are you on a Linux or Window?
Did you install python? ( https://www.python.org/downloads/ )
Under Windows, the PATH environment variable need to be updated to include the python directory.

Following is the exact commands I running.

C:
set PYTHONPATH=C:\Users\mveillette\AppData\Local\Programs\Python\Python35-32\Lib;C:\Users\mveillette\Documents\git\pyang
python.exe C:\Users\mveillette\Documents\git\pyang\bin\pyang --sid-help
python.exe C:\Users\mveillette\Documents\git\pyang\bin\pyang --list-sid --generate-sid-file 1500:100 ietf-interfaces@2014-05-08.yang<mailto:ietf-interfaces@2014-05-08.yang>

Path C:\Users\mveillette\Documents\git and C:\Users\mveillette\AppData need to be adjusted to your environment.

Regards,
Michel

**Comment by gruber6121** [comment]
I am trying to run it at the same time on Linux and Windows. 
For Windows it already works. Thank you for your support.

**Comment by lemikev** [comment]
Just curious
Who you are?

**Comment by lemikev** [comment]
COMMENT #1
I would prefer the following options:
-f sid --sid-range entry-point:size
-f sid --sid-update-from

ANSWER
We'll like to keep the current appraoch for the following reasons:
The initial version of this plugin was based on the -f option but this approach has multiple drawbacks. When -f is used, the module name need to be repeated for the .sid file which is more work and error prone.  Furthermore, the different options (i.e. generate, update and check) have two outputs, the stdout for the information to the end user and a file for the .sid creation. This dual output is not supported by the -f option.
Finally, changing these options will impact the https://comi.space/ web site. 

=================================================================

COMMENT #2
Please also document your options in doc/pyang.1.dbk.

ANSWER
doc/pyang.1.dbk have been updated.

=================================================================

COMMENT #3
Please add some tests, create a new directory test/test_sid; do positive and negative testing

ANSWER
Seven tests have been added.

=================================================================

COMMENT #4
(e.g., I tried to reference a non-existing file with --update-sid-file, and it gave me a stacktrace)

ANSWER
I'm unable to reproduce this error with the current version.

=================================================================

COMMENT #5
@@ -0,0 +1,566 @@
+ """sid plugin
+ 
+  Plugin used to generate or update .sid files.
  
Please include a reference to the relevant specs.

ANSWER
The following text have been added:

"""sid plugin

Plugin used to generate or update .sid files.
Please refer to [I-D.ietf-core-sid], [I-D.ietf-core-comi], [I-D.ietf-core-yang-cbor]
and [I-D.ietf-core-yang-library] for more information.

"""

=================================================================

COMMENT #6
+     def __init__(self, msg=""):
+         self.msg = msg
+ 
+  class SidParcingError(Exception):
 
This should be SidParsingError

ANSWER
Fixed

=================================================================

COMMENT #7
+ 
+    ########################################################
+    # Retrieve the module revision from the pyang context
+    def get_module_revision(self, module):

Please use util.get_latest_revision(module)

ANSWER
Done

**Comment by mbj4668** [comment]
Hi,

Thanks for addressing my comments.  I think it's great to get this plugin built-in into pyang. 

> COMMENT #1
> I would prefer the following options:
> -f sid --sid-range entry-point:size
> -f sid --sid-update-from
> 
> ANSWER
> We'll like to keep the current appraoch for the following reasons:
> The initial version of this plugin was based on the -f option but this approach has multiple drawbacks. When -f is used, the module name need to be repeated for the .sid file which is more work and error prone. Furthermore, the different options (i.e. generate, update and check) have two outputs, the stdout for the information to the end user and a file for the .sid creation. This dual output is not supported by the -f option.

Ok, fair enough.  I still would have prefered -f for consistency, and I think it would have worked fine, e.g. additional info can be printed to stderr and/or the -o could have been used, just like for other formats.  But I think it is more important to get this plugin into the main branch.


However, the convention in pyang's options is to use the name of the plugin as a prefix, e.g.:

    Name output specific options:
      --name-print-revision
                        Print the name and revision in name@revision format
    Tree output specific options:
      --tree-help         Print help on tree symbols and exit
      --tree-depth=TREE_DEPTH
                        Number of levels to print
  
(do pyang --help) to the all options.

So I think you should change "--list-sid" to "--sid--list" and "--generate-sid-file" to "--sid-generate-file", 
"count" to "--sid-count" etc.

File: doc/pyang.1.dbk [hunk 5]
```diff
@@ -937,6 +971,108 @@
...
+      in constrained applications such as COREconf. This plugin is used to
+      automatically generate and updated .sid files used to persist and
+      distribute SID assignments.
+    </para>
+    <para>
+      Options for generating, updating and checking .sid files:
+    </para>
+    <variablelist>
+      <varlistentry>
+        <term><option>--generate-sid-file</option></term>
```
Comment (mbj4668) [inline doc/pyang.1.dbk]: you should document the arguments to this option.  I tried "pyang --generate-sid-file foo,yang" and it just hangs.

File: doc/pyang.1.dbk [hunk 6]
```diff
@@ -937,6 +971,108 @@
...
+            This option is used to generate a new .sid file from a YANG module. This example
+            shows how to generate the file <emphasis>toaster@2009-11-20.sid</emphasis>.
+          </para>
+          <informalexample>
+            <screen>pyang --generate-sid-file 20000:100 toaster@2009-11-20.yang</screen>
+          </informalexample>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><option>--update-sid-file</option></term>
```
Comment (mbj4668) [inline doc/pyang.1.dbk]: document the argument in the <option> i.e.  <replaceable>prevsidfile<replaceable>

File: doc/pyang.1.dbk [hunk 7]
```diff
@@ -937,6 +971,108 @@
...
+            how to generate the file <emphasis>toaster@2009-12-28.sid</emphasis> based
+            on the SIDs already present in <emphasis>toaster@2009-11-20.sid</emphasis>.
+          </para>
+          <informalexample>
+            <screen>pyang --update-sid-file toaster@2009-11-20.sid toaster@2009-12-28.yang</screen>
+          </informalexample>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><option>--check-sid-file</option></term>
```
Comment (mbj4668) [inline doc/pyang.1.dbk]: document the argument to the option

File: doc/pyang.1.dbk [hunk 8]
```diff
@@ -937,6 +971,108 @@
...
+            using the initial range(s) present in <emphasis>toaster@2009-11-20.sid</emphasis>
+            and the extra range specified in the command line.
+          </para>
+          <informalexample>
+            <screen>pyang --update-sid-file toaster@2009-11-20.sid toaster@2009-12-28.yang --extra-sid-range 20100:100</screen>
+          </informalexample>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><option>count</option></term>
```
Comment (mbj4668) [inline doc/pyang.1.dbk]: this is not an option, but an argument to --generate-sid-file.  However, if I understand this correctly, if you give this argument to --generate-sid-file it doesn't actually generate any file, which seems weird.

File: doc/pyang.1.dbk (hunk 5 above)
Comment (lemikev) [reply doc/pyang.1.dbk]: pyang hangs on the following stdin.read().
If you quickly understand why, please send my a propose fix.

    if o.hello:
        ctx.capabilities = hel.registered_capabilities()
        for (mn,rev) in hel.yang_modules():
            mod = ctx.search_module(0, mn, rev)
            if mod is None:
                emarg = mn
                if rev: emarg += "@" + rev
                sys.stderr.write(
                    "module '%s' specified in hello not found.\n" % emarg)
                sys.exit(1)
            modules.append(mod)
    else:
        if len(filenames) == 0:
            text = sys.stdin.read()

Comment (mbj4668) [reply doc/pyang.1.dbk]: My point is that the documentation seems to indicate that it is ok to pass just "--generate-sid-file" <filename>, but the code will treat <filename> as an argument to "--generate-sid-file", and then wait for an input module from stdin.

So the behavior is as expected, but the documentation is not.

**Comment by lemikev** [comment]
The updated man page is available for review. The sid.py update for the new option names and associated test files will follow.

**Comment by lemikev** [comment]
Negative tests have been added.
Following is the answerw to the latest set of comments.

COMMENT #8

So I think you should change "--sid-list" to "--sid--list" and "--generate-sid-file" to "--sid-generate-file",
"count" to "--sid-count" etc.

ANSWER
Done

COMMENT #9
--generate-sid-file
you should document the arguments to this option. I tried "pyang --generate-sid-file foo,yang" and it just hangs.

ANSWER
The SYNOPSIS have been fixed, the required arguments have been documented.

COMMENT #10
--update-sid-file
document the argument in the i.e. prevsidfile

ANSWER
The required arguments have been documented.

COMMENT #11
--check-sid-file
document the argument to the option

ANSWER
The required arguments have been documented.

COMMENT #12
count
this is not an option, but an argument to --generate-sid-file. However, if I understand this correctly, if you give this argument to --generate-sid-file it doesn't actually generate any file, which seems weird.

ANSWER
Tag <option> have been replaced by <emphasis>

The count argument is available for both the --generate-sid-file and the --update-sid-file options. When specified, the generation or update process is performed as usual but instead of assigning SIDs, the number of SIDs required is computed. This option have been introduced during the development of the https://comi.space/ web site to automatically propose to the end user the next entry point and a reasonable range size when generating or updating a .sid file.

File: doc/pyang.1.dbk [hunk 9]
```diff
@@ -937,6 +971,147 @@
...
+            For example:
+          </para>
+          <informalexample>
+            <screen>$ pyang --sid-check-file toaster@2009-12-28.sid \
+toaster@2009-12-28.yang</screen>
+          </informalexample>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><option>--list_sid</option></term>
```
Comment (mbj4668) [inline doc/pyang.1.dbk]: this should be --sid-list (here and in a few other places)

**Comment by lemikev** [comment]
I'm done on my side unless you have more comments for me.
